pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java

```diff
@@ -231,7 +232,7 @@ public long getCommittedTxnCount() {
     @Override
     public CompletableFuture<Position> appendBufferToTxn(TxnID txnId, long sequenceId, ByteBuf buffer) {
         buffer.retain();
-        return transactionBufferFuture.thenCompose(ignore -> {
+        CompletableFuture<Position> future = publishFuture.thenCompose(ignore -> {
```

Comment:
https://github.com/apache/pulsar/blob/99e6aebb88c8ea82c8b8d65261ce71f0ab8d36d8/pulsar-common/src/main/java/org/apache/pulsar/common/util/FutureUtil.java#L174 

The Sequencer can be used to handle this case. So that you don't need to create a meaningless `publishFuture` which is not easy to understand.

For example:

```java
    public void testSequencer2() {
        final FutureUtil.Sequencer<Void> sequencer = FutureUtil.Sequencer.create();
        CompletableFuture<Void> first = new CompletableFuture<>();
        sequencer.sequential(() -> first);
        sequencer.sequential(() -> CompletableFuture.runAsync(() -> System.out.println("1")));
        sequencer.sequential(() -> CompletableFuture.runAsync(() -> System.out.println("2")));
        sequencer.sequential(() -> CompletableFuture.runAsync(() -> System.out.println("3")));
        first.complete(null);
    }
```

It will consistently execute as:

```
1
2
3
```

Reply:
The actual situation is different from this example.
We have a global variable `transactionBufferFuture` whose generic type is Void.
```java
private final CompletableFuture<Void> transactionBufferFuture = new CompletableFuture<>();
```
But the method generic type of the method `appendBufferToTxn` is Position.
```java
public CompletableFuture<Position> appendBufferToTxn(TxnID txnId, long sequenceId, ByteBuf buffer) 
```
So, we inevitably introduce a new future within this method. And since `sequencer.sequential` involves thread-switching operations, we need to perform an additional `buffer.retain();`. Later, if there is any optimization for `sequencer.sequential`, we also need to consider its impact on `buffer.retain();`. This will introduce some complexity.
```
  public CompletableFuture<Position> appendBufferToTxn(TxnID txnId, long sequenceId, ByteBuf buffer) {
      buffer.retain();
      final CompletableFuture<Position> publishFuture = new CompletableFuture<>();
      sequencer.sequential(() -> {
          buffer.retain();
          return transactionBufferFuture.thenCompose(ignore -> {})
                  .thenAccept(publishFuture::complete);
      }).exceptionally(ex -> {
          publishFuture.completeExceptionally(ex);
          return null;
      });
      return publishFuture;
  }
```

Reply:
Or we can modify the paradigm of `transactionBufferFuture` so that it can adapt to `sequencer` in this method.
```java
private final CompletableFuture<Position> transactionBufferFuture = new CompletableFuture<>();
```
But this generic type position is not used. It is also useless and difficult to understand.
Moreover, it also has the problem of `buffer.retain()`;